ext/pg_text_decoder.c

```diff
@@ -392,6 +408,180 @@ pg_text_dec_from_base64(t_pg_coder *conv, char *val, int len, int tuple, int fie
+}
+
+static int str4_to_int(const char *str)
+{
+  return char_to_digit(str[0]) * 1000
+       + char_to_digit(str[1]) * 100
+       + char_to_digit(str[2]) * 10
+       + char_to_digit(str[3]);
+}
+
+static int str2_to_int(const char *str)
+{
+  return char_to_digit(str[0]) * 10
+       + char_to_digit(str[1]);
+}
+
+static VALUE pg_text_decoder_timestamp_do(VALUE rstr, int with_timezone)
+{
+  const char *str = StringValuePtr(rstr);
+
+  if (isdigit(str[0]) && isdigit(str[1]) && isdigit(str[2]) && isdigit(str[3])
+   && str[4] == '-'
+   && isdigit(str[5]) && isdigit(str[6])
+   && str[7] == '-'
+   && isdigit(str[8]) && isdigit(str[9])
+   && str[10] == ' '
+   && isdigit(str[11]) && isdigit(str[12])
+   && str[13] == ':'
+   && isdigit(str[14]) && isdigit(str[15])
+   && str[16] == ':'
+   && isdigit(str[17]) && isdigit(str[18])
+  )
+  {
+    struct tm tm;
+    int nsec = 0;
+    int tz_neg = 0;
+    int tz_hour = 0;
+    int tz_min = 0;
+    int tz_sec = 0;
+ 
+    tm.tm_year = str4_to_int(&str[0]) - 1900;
+    tm.tm_mon = str2_to_int(&str[5]) - 1;
+    tm.tm_mday = str2_to_int(&str[8]);
+    tm.tm_hour = str2_to_int(&str[11]);
+    tm.tm_min = str2_to_int(&str[14]);
+    tm.tm_sec = str2_to_int(&str[17]);
+    tm.tm_isdst = 0;
+    str += 19;
+ 
+    if (str[0] == '.' && isdigit(str[1]))
+    {
+      /* nano second part, up to 9 digits */
+      static const int coef[9] = {
+        100000000, 10000000, 1000000,
+        100000, 10000, 1000, 100, 10, 1
+      };
+      int i;
+
+      str++;
+      for (i = 0; i < 9 && isdigit(*str); i++)
+      {
+        nsec += coef[i] * char_to_digit(*str++);
+      }
+    }
+
+    if (with_timezone)
+    {
+      if ((str[0] == '+' || str[0] == '-') && isdigit(str[1]) && isdigit(str[2]))
+      {
+        tz_neg = str[0] == '-';
+        tz_hour = str2_to_int(&str[1]);
+        str += 3;
+      }
+      if (str[0] == ':')
+      {
+        str++;
+      }
+      if (isdigit(str[0]) && isdigit(str[1]))
+      {
+        tz_min = str2_to_int(str);
+        str += 2;
+      }
+      if (str[0] == ':')
+      {
+        str++;
+      }
+      if (isdigit(str[0]) && isdigit(str[1]))
+      {
+        tz_sec = str2_to_int(str);
+        str += 2;
+      }
+    }
+    if (*str != '\0')
+    {
+      // not consumed all the string
+      return rstr;
+    }
+
+    if (with_timezone)
+    {
+#ifdef _WIN32
+      /* we can't use _mkgmtime because it is not available when using mingw32 */
+      time_t time;
+      time_t prevTZ = _timezone;
+      _timezone = 0;
+      time = mktime(&tm);
+      _timezone = prevTZ;
+#else
+      time_t time = timegm(&tm);
+#endif
+      if (time != -1)
+      {
+        struct timespec ts;
+        int gmt_offset;
+
+        gmt_offset = tz_hour * 3600 + tz_min * 60 + tz_sec;
+        if (tz_neg)
+        {
+          gmt_offset = - gmt_offset;
+        }
+        ts.tv_sec = time - gmt_offset;
+        ts.tv_nsec = nsec;
+        return rb_time_timespec_new(&ts, gmt_offset);
```

Comment:
Couldn't we use `rb_funcall()` to Time.new as fallback for Ruby<2.3? This way the Regexp version could be removed entirely.